src/node/interfaces.cpp

```diff
@@ -940,19 +940,12 @@ class MinerImpl : public Mining
 
     BlockRef waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
     {
-        // Interrupt check interval
-        const MillisecondsDouble tick{1000};
-        auto now{std::chrono::steady_clock::now()};
-        auto deadline = now + timeout;
-        // std::chrono does not check against overflow
-        if (deadline < now) deadline = std::chrono::steady_clock::time_point::max();
+        if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
-            while ((notifications().m_tip_block == uint256() || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
-                now = std::chrono::steady_clock::now();
-                if (now >= deadline) break;
-                notifications().m_tip_block_cv.wait_until(lock, std::min(deadline, now + tick));
-            }
+            notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                return (notifications().m_tip_block != current_tip && notifications().m_tip_block != uint256::ZERO) || chainman().m_interrupt;
```

Comment:
I think that since `waitTipChanged` takes a `current_tip` argument the `m_tip_block != uint256::ZERO` check is no longer needed. Though it seems fine to keep it.

IIRC the reason I added it was in case `waitTipChanged` was called during early node initialization, before `m_tip_block` is set to the genesis block.

That scenario is now handled by `m_tip_block != current_tip`.

In term of how the Template Provider works, it used to rely on this check, but now it just waits for `getTip()` to return a value before calling `waitTipChanged()`. .

When implemented as part of Bitcoin Core, as in #29432, it can't call `waitTipChanged()` earlier, because it has no way of knowing what to set for `current_tip`.

When implemented as an external application (e.g. https://github.com/Sjors/bitcoin/pull/48) it might know the tip from some other source before `m_tip_block` is set. So it could call `waitTipChanged()` very early in node initialization, but that still doesn't require the `m_tip_block != uint256::ZERO` check.

Reply:
Either approach seems ok. I probably wouldn't have added the null check, but it does a provide a small convenience if miner is running while the node is restarted, since waitTipChanged will patiently wait for startup to complete and a new block to be connected, instead of returning 0 on the first waitTipChanged call and requiring the miner to make a second call.

Whether we keep this check or drop it, it would probably be good to update the `waitTipChanged` documentation for the method to say that it either returns 0 on startup (if null check is dropped) or waits for the first block to be connected on startup (if null check is kept).

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30967#discussion_r1778162781

> In terms of how the Template Provider works, it used to rely on this check, but now it just waits for `getTip()` to return a value before calling `waitTipChanged()`. .
> 
> When implemented as part of Bitcoin Core, as in #29432, it can't call `waitTipChanged()` earlier, because it has no way of knowing what to set for `current_tip`.

I may be misunderstanding, but this doesn't seem right to me. If a caller just wants to know when a new block is connected it shouldn't need to call  `getTip()` in a loop until it returns non-null before calling `waitTipChanged()`, if that is what this comment is saying. The code should just be able to call `waitTipChanged()` with the zero value, which will return as soon as node is started and it knows what the tip is. (But I do agree with Marco it might be better to make let waitTipChanged() take a std::optional hash and not treat 0 as special). If this doesn't work, it suggests there could be a bug and code is not setting `m_tip_block` reliably.

Reply:
> The code should just be able to call `waitTipChanged()` with the zero value, which will return as soon as node is started and it knows what the tip is.

That indeed seems like a better approach, will change that for the TP.